### ISSUE TITLE
[FEATURE] - Save GitHub Codespaces projects always as "remote project"

### ISSUE THREAD

**Issue by TylerLeonhardt** [alefragnani/vscode-project-manager#479]
First off, this extension is AMAZING! Thank you for all that you do to make this extension a reality. The support for Remote Development is really nice:

![image](https://user-images.githubusercontent.com/2644648/112576817-49644280-8db0-11eb-8294-c7157c679269.png)

Some of those remotes are GitHub Codespaces, some are Remote - Containers 😄 

This issue _only happens in GitHub Codespaces in the browser_.

**Environment/version**

- Extension version: 12.0.1
- VSCode version: 1.55-insiders web (Codespaces)
- OS version: macOS client/Linux server

**Steps to reproduce**

1. Open a repo on GitHub (like https://github.com/Microsoft/vscode )
2. Open this repo in GitHub Codespaces
3. Install Project Manager
4. Save this folder

Expected:

It saves as the Remote url like:
```
{
		"name": "vscode",
		"rootPath": "vscode-remote://codespaces+tylerleonhardt-microsoft-vscode-4ds4/home/node/workspace/vscode",
		"paths": [],
		"group": "",
		"enabled": true
}
```

Actual:

```
{
		"name": "vscode",
		"rootPath": "/home/node/workspace/vscode",
		"paths": [],
		"group": "",
		"enabled": true
}
```

*Important*

When I use VS Code _client_ to connect to this Codespace (GitHub Codespaces isn't just available in the browser, it's also available via the desktop client), it does save properly (which is how I got the value for the `Expected` result)

**Comment by alefragnani** [comment]
Hi @TylerLeonhardt ,

First of all, thank you for you kind words! I'm glad to know the extension is being helpful. 👍 

About the issue, this is how the extension is supposed to work. This behavior was based on user's feedback after the initial release which supported remotes. So, I created #323 to handle those changes on how the extension should work when installed locally or remotely.  

TL;DR, you can save any _remote project_ on your local installation and be able to open it. it automatically connects to the  _remote_. Otherwise, if you want your remote projects available when you are connected to that remote location, you also can.

But I would like to understand your scenario. Which is your the expected behavior when you are connected to Codespaces (on browser) and save a project? 
* Would you like Codespaces to be opened _"on that new location/folder"_ instead of the root folder of your Codespace? 
* Would you like the Codespace project/url to be available to your _VS Code client_, like Settings Sync?

I'm not sure how the Codespaces extension work (didn't used yet) but if it has the same _core_ as the SSH/WSL/Docker extensions, I suppose there is no API to _open some specific location on the remote_, and that was one of the reasons I couldn't allow saving _remote locations_ when the extension is installed locally.

I'm using remotes a bit more (mostly Docker and Codespaces on browser), so I'm interested on improving remote experience.

Thank you

**Comment by TylerLeonhardt** [comment]
I think I'm not articulating the issue I'm seeing correctly. Let me first say that Project Manager already works with the Remote Codespaces extension for VS Code _client_. However, Project Manager isn't working like I would expect from Codespaces on the Web.

Let me elaborate.

From my perspective, when I use Codespaces, I don't have many projects in the same Codespace... I have 1 Codespace _per project_ (in my case, per repo)

Example:

* I go to github.com/tyler/project-1
* open repo in Codespaces
* **In a different tab** go to github.com/tyler/project-2
* open repo in Codespaces

These are 2 completely different machines. What I would expect is, if I click on a favorite, then I would be redirected to that codespace... or If I click the "open in a new window" it opens in a new tab in my browser.

Actually what's amazing about this is that this is exactly how it works today, if you provide Project Manager the following config:

```
{
		"name": "vscode",
		"rootPath": "vscode-remote://codespaces+tylerleonhardt-microsoft-vscode-4ds4/home/node/workspace/vscode",
		"paths": [],
		"group": "",
		"enabled": true
}
```

where `rootPath` is a remote path.

However, the problem is _saving_ a project. Locally, a Codespace gets saved like that ^ but with Project Manager inside Codespaces on the web, it gets saved like:

```
{
		"name": "vscode",
		"rootPath": "/home/node/workspace/vscode",
		"paths": [],
		"group": "",
		"enabled": true
}
```

which is not ideal because then the association that it is in codespaces is gone.

**Comment by TylerLeonhardt** [comment]
> Would you like the Codespace project/url to be available to your VS Code client, like Settings Sync?

Yes there should be some concept of syncing projects... (which you already have an issue about https://github.com/alefragnani/vscode-project-manager/issues/431) but what I find interesting is that for this scenario, Codespaces Web will likely only open other Codespaces. Other remotes (like ssh, containers, etc) are not useful in Codespaces Web because those other remotes aren't available.

**Comment by alefragnani** [comment]
No problem, I think we both aren't articulating 😆 . But with your detailed scenario I could confirm what I thought to be your request. You would like that:

_Even if I install the extension on remote (Codespaces) the saved/favorite project should point to its remote location_

I understand, and agree, that Codespaces (web) should work this way. It would be very handy.

I'll take a look if this can be done with the current APIs, because I remember when I was adding the Remote support, that VS Code didn't exposed (or I didn't find) the _remote location_ when the extension was installed _on remote_.

Hope this helps

**Comment by alefragnani** [comment]
> Codespaces Web will likely only open other Codespaces

I wonder if Codespaces Web shouldn't open VS Code _client_ as well. Really handy integration.

**Comment by TylerLeonhardt** [comment]
> I'll take a look if this can be done with the current APIs, because I remember when I was adding the Remote support, that VS Code didn't exposed (or I didn't find) the remote location when the extension was installed on remote.

Interesting... yeah let me know how that goes. If you don't find anything obvious I can ask my team.

> I wonder if Codespaces Web shouldn't open VS Code client as well. Really handy integration.

ooo great point. If it's a local file path or Remote ssh or Remote container, it could open in VS Code _client_. That'll get a bit tricky because the user might use VS Code Insiders vs VS Code stable... but you could probably have two buttons for that or a quick pick to say Insiders or not.

**Comment by alefragnani** [comment]
> Interesting... yeah let me know how that goes. If you don't find anything obvious I can ask my team.

Thanks, that would be awesome 👍 

This is what I can get while `debuging` on Codespaces web.

```shell
vscode.env.remoteName
'codespaces'
vscode.env.sessionId
'da10adb8-some-random-guid-value-a30871616883191168'
vscode.env.machineId
'7d14f8-some-other-random-guid-value-964cd1036'
vscode.env.remoteName
'codespaces'
vscode.workspace.workspaceFolders
(1) [{…}]
vscode.workspace.workspaceFolders[0]
{uri: h, name: 'vscode-separators', index: 0}
index:0
name:'vscode-separators'
uri:h {scheme: 'file', authority: '', path: '/home/codespace/workspace/vscode-separators', query: '', fragment: '', …}
_formatted:'file:///home/codespace/workspace/vscode-separators'
_fsPath:'/home/codespace/workspace/vscode-separators'
authority:''
fragment:''
fsPath (get):ƒ fsPath(){return this._fsPath||(this._fsPath=c(this,!1)),this._fsPath}
path:'/home/codespace/workspace/vscode-separators'
query:''
scheme:'file'
__proto__:g
__proto__:Object
vscode.env.appRoot
'/home/codespace/.vscode-remote/bin/d06d2fsomerandomvalues1a9087d225173e'
vscode.env.appName
'Visual Studio Code - Insiders'
```

It's curious to see that `vscode.env.appName` returns `Visual Studio Code - Insiders`, but `vscode.env.remoteName` returns `codespaces` and `vscode.env.appRoot` is `/home/codespaces/.vscode-remote/...`

And, look that the `scheme` returns `file`, instead of `vscode-remote` or `codespaces`, and  `vscode.workspace.workspaceFolders[0]` returns `/home/codespace/workspace/vscode-separators`, which is a _local_ instead of _Codespaces_ like Uri.

I wonder if there are other ways to detect/identify the remote Uri, while running on remote.

> ooo great point. If it's a local file path or Remote ssh or Remote container, it could open in VS Code client. That'll get a bit tricky because the user might use VS Code Insiders vs VS Code stable... but you could probably have two buttons for that or a quick pick to say Insiders or not.

Well, Codespaces web already has a `Open in Visual Studio Code` button, even if in my case, I only turned on setting sync to the `Insiders` release (I would say it's a bug then). Anyway, Codespaces already has some kind of _connection_ to VS Code _client_ 😬 

Thank you

**Comment by alefragnani** [comment]
Maybe this is not a VS Code issue, but Codespaces. 

I mean, maybe the Codespaces extension should export an API with such information. The same for the other Remote extensions (Docker, SSH, WSL).

**Comment by TylerLeonhardt** [comment]
Looking through vscode's codebase, I found this:

https://github.com/microsoft/vscode/blob/99c8eb699429785a11e02090b54de275b498991f/extensions/github/src/pushErrorHandler.ts#L35-L37

might be useful :)

**Comment by alefragnani** [comment]
Yep, I guess 😬 .

I mean, It has a bunch of information, and the `Uri` is not there, but if you combine a few pieces, it has the same results (at least looking two of my workspaces).

In the end, I had to do: `vscode-remote://codespaces+${info.name}${vscode.workspaceFolders[0].uri.fsPath}`

But, I don't know how you did use the extension _installed on Codespaces_. I mean, when I installed, the projects are only available on the same Codespace. When I open another Codespace, the list it's empty. And in fact, this is the expected behavior because the projects are not stored on any location/area that Settings Sync watches. In Codespaces, it's being saved on `/home/codespace/.vscode-remote/data/User/globalStorage/alefragnani.project-manager/projects.json`, but even being the same _path_ on the other Codespace, it's not the same file.

Could you confirm my assumption about a _manipulated Uri_ is the correct approach for this?

Thank you

**Comment by alefragnani** [comment]
BTW, I just find out that debugging a VS Code extension on Codespaces (web and desktop) isn't working properly. 

## Web

When you start the debugging session, the new `[Extension Development Host]` is empty (expected behavior), but if you select/open any other Codespace (because I need to open a valid workspace to test some extension), the original debugging session does not recognizes it, and it _freezes_. The debugging toolbar is there, but no debugging at all. 

## Local

The same startup scenario, but instead of _frezzing_ the debugging session, it simply stops. The debugging toolbar closes but new `[Extension Development Host]` remains. But in this case, no matter if I select a Codespace or a local folder, it happens the same

I'm not sure you did experience this, or I'm doing something wrong 😕 .

**Comment by TylerLeonhardt** [comment]
> But, I don't know how you did use the extension installed on Codespaces. I mean, when I installed, the projects are only available on the same Codespace.

Right so what you describe is what I see. I was simply copying and pasting the `projects.json` file between my client and Codespaces web. Since I use the same few Codespace containers, copying and pasting the `projects.json` file isn't that bad as an interim until a "sync" feature is supported.

BTW, on that topic, the settings sync extension before the in-box settings sync used a github gist to persist `projects.json` files. Perhaps that is a way to go until Settings Sync properly supports your scenario. (We can probably have this chat in that other issue though).

**Comment by TylerLeonhardt** [comment]
> Could you confirm my assumption about a manipulated Uri is the correct approach for this?

Yeah this is probably the right way to go for now. If I hear of a better way, I'll let you know.

> but if you select/open any other Codespace (because I need to open a valid workspace to test some extension), the original debugging session does not recognizes it, and it freezes

Interesting. Yeah I think Web should gracefully stop instead in that case. Can you open an issue about that in https://github.com/Microsoft/vscode ? I bet you can do a very simple repo by creating a simple extension that calls whatever API you're calling to open the Codespace or local folder.

**Comment by alefragnani** [comment]
> Yeah this is probably the right way to go for now. If I hear of a better way, I'll let you know.

That’s great! I’ll use this approach then. Thanks!

Also, I created https://github.com/microsoft/vscode/issues/120142, asking for a way to use Settings Sync to sync the `projects.json` file. I thought there was already an issue about that, but didn’t find any.

> Interesting. Yeah I think Web should gracefully stop instead in that case. Can you open an issue about that

Sure. I’ll prepare a sample repo and make sure it can be reproduced. I dare to say it won’t be necessary to use the `openFolder` command. I guess even the `File / Open` menu will have the same behavior.

Thank you

**Comment by alefragnani** [comment]
When you save a _remote_ project, it uses the `remote-explorer` codicon, no matter the kind of remote (Docker, SSH, WSL and Codespaces). What about using a different codicon for Codespaces? I was thinking about `GitHub` 😬 

> BTW, on that topic, the settings sync extension before the in-box settings sync used a github gist to persist projects.json files. Perhaps that is a way to go until Settings Sync properly supports your scenario. (We can probably have this chat in that other issue though).

Sure, feel free to add comments/suggestions there. I was thinking about it today and two different ideas came to mind, so would be great to have your feedback/experience on this. 

Thank you

**Comment by TylerLeonhardt** [comment]
> I was thinking about GitHub 😬

Make sure you follow https://github.com/logos (in my personal opinion in words that are my own and don't represent my employer 😄 , I think `github` is a good move here since you are opening a GitHub Codespace)

Ideally Docker, SSH, WSL, Codespaces all have their own icons but you're limited to what you have so I think this is a good first step.

**Comment by alefragnani** [comment]
Yep, there is no specific codicon for each remote right now, but maybe @misolori will add them in the future. Who knows 😆

**Comment by TylerLeonhardt** [comment]
Maybe for now the tool tips can say what kind of remote they are and any interesting info you can get?

**Comment by alefragnani** [comment]
Good point! 

Today it only displays the project's path (not so relevant for remotes, tbh). I can extract its kind and I guess the relative `Uri` too. 

I wonder if would be doable to detect its _status_. 

Thanks

**Comment by misolori** [comment]
For existing codicons you could use:

**Codespaces**
![image](https://user-images.githubusercontent.com/35271042/113225345-42bb4c80-9242-11eb-98aa-cd6af99d5a4a.png)

**Container**
![image](https://user-images.githubusercontent.com/35271042/113225366-5070d200-9242-11eb-9046-82a5af2ae146.png)

**SSH**
![image](https://user-images.githubusercontent.com/35271042/113225397-61b9de80-9242-11eb-8669-5ac3be493274.png)

**WSL**
Don't have a penguin in here though we could add one

**Comment by alefragnani** [comment]
Great idea @misolori ! Thank you 👍

**Comment by alefragnani** [comment]
It works pretty well..

![Screen Shot 2021-03-31 at 21 25 00](https://user-images.githubusercontent.com/3781424/113226888-a9069600-9267-11eb-8710-1e4672a07287.png)


There is one small detail. I have to deal with the icon's default color, making it to use the _system_ color. 

```ts
 new ThemeIcon('symbol-method', new ThemeColor("icon.foreground"));
```

Created #483 to track this

**Comment by eamodio** [comment]
Not sure this will work, but what does `await vscode.env.asExternalUri(vscode.workspace.workspaceFolders[0].uri)` return? In theory I would expect it to return the codespace uri.

**Comment by alefragnani** [comment]
It raises an error 😕 . If I understand correctly, `asExternalUri` doesn't supports a `file` scheme, only _externals_. By the docs

```
         * Resolves a uri to form that is accessible externally. Currently only supports `https:`, `http:` and
         * `vscode.env.uriScheme` uris.
```

**Comment by eamodio** [comment]
Ah, and if you do `vscode.env.asExternalUri(vscode.workspace.workspaceFolders[0].uri.with({ scheme: 'codespace' })`? (the scheme might be `codespaces` I can't remember for sure.

**Comment by alefragnani** [comment]
Didn't work. Neither using `vscode.env.asExternalUri(vscode.workspace.workspaceFolders[0].uri.with({ scheme: vscode.env.uriScheme })`, which in this case is `vscode-insiders`.

In the meantime, the other approach (combining pieces) seems to work, but I agree this one is much more elegant.

**Comment by TylerLeonhardt** [comment]
👀

**Comment by alefragnani** [comment]
Just released the update. Could you verify if it works as expected?

**Comment by TylerLeonhardt** [comment]
Works so well! Now I need to open a feature request to have non-Codespaces projects open in local VS Code or VS Code insiders :)

**Comment by TylerLeonhardt** [comment]
proof: 
https://user-images.githubusercontent.com/2644648/113449095-f727b080-93b1-11eb-9525-841a80eb0c86.mp4

**Comment by alefragnani** [comment]
I thought about that too 👍

It would fit perfectly with the Settings Sync feature.

**Comment by misolori** [comment]
@alefragnani FYI, we now have a penguin (`terminal-linux`) codicon 😉 

![image](https://user-images.githubusercontent.com/35271042/113775135-1dfd2400-96dd-11eb-806b-056fb3fdd201.png)

**Comment by alefragnani** [comment]
@misolori thank you! It will fit perfectly 🤝 

Updated #483 to describe all codicons are available
